airflow/sensors/external_task_sensor.py

```diff
@@ -16,22 +16,71 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
-
 import datetime
 import os
-from typing import Optional, Union
+from typing import FrozenSet, Optional, Union
 
 from sqlalchemy import func
 
+from airflow.configuration import conf
 from airflow.exceptions import AirflowException
-from airflow.models import DagBag, DagModel, DagRun, TaskInstance
+from airflow.models import BaseOperatorLink, DagBag, DagModel, DagRun, TaskInstance
 from airflow.operators.dummy_operator import DummyOperator
 from airflow.sensors.base_sensor_operator import BaseSensorOperator
 from airflow.utils.decorators import apply_defaults
 from airflow.utils.session import provide_session
 from airflow.utils.state import State
 
 
+def get_possible_target_execution_dates(execution_date, execution_delta, execution_date_fn):
+    """
+    Gets the execution date(s) of an external DAG for which an
+    ExternalTaskSensor should succeed on. Default is the execution
+    date itself, but it may be modified if a non-null execution delta
+    or execution date function is passed in.
+
+    :param execution_date: The execution date of the sensor
+    :type execution_date: datetime.datetime
+    :param execution_delta: Time difference between the sensor
+        execution date and the target DAG run execution date. Positive
+        delta looks back in time.
+    :type execution_delta: Optional[datetime.timedelta]
+    :param execution_date_fn: Function to compute the execution date(s)
+        of the target DAG run to look at given the sensor's execution
+        date.
+    :type execution_date_fn: Optional[Callable]
+    :return: Execution date(s) to wait for
+    :rtype: List[datetime.datetime]
+    """
+    if execution_delta:
+        dttm = execution_date - execution_delta
+    elif execution_date_fn:
+        dttm = execution_date_fn(execution_date)
+    else:
+        dttm = execution_date
+
+    return dttm if isinstance(dttm, list) else [dttm]
+
+
+class ExternalTaskLink(BaseOperatorLink):
+    name = 'External DAG'
+
+    def get_link(self, operator, dttm):
+        possible_execution_dates = get_possible_target_execution_dates(
+            execution_date=dttm,
+            execution_delta=getattr(operator, 'execution_delta', None),
+            execution_date_fn=None,
```

Comment:
`execution_date_fn` is always `None`? Does this mean `ExternalTaskLink` won't work with `ExternalTaskSensor` that defines `execution_date_fn` ?

Reply:
That's correct. I don't think we're able to serialize / deserialize functions, so that would be incompatible with Airflow 2.0

Reply:
Oh that's unfortunate if we can't make an existing `execution_date_fn` feature work here because I assume many people would need to be using `execution_date_fn` to shift execution_date around according to some fancy logic they may have.

I have one suggestion that might work. If we know the `task_id` and `dag_id` of the `operator`, we can always find the real `ExternalTaskSensor` object (e.g. by looking it up in the `DagBag`). Once we have a reference to the real `ExternalTaskSensor` object, we can call its `execution_date_fn`.

If that doesn't work for you, I think it's important to make sure that if the `ExternalTaskSensor` uses `execution_date_fn`, the extra link doesn't get generated. In the current state of the PR, it looks like the link is going to point to the wrong `execution_date` if the user intended to use `execution_date_fn`. That is going to cause confusion. Also a comment that says `execution_date_fn` is not supported for the extra link is probably helpful too.

Reply:
@kaxil Could you confirm whether this approach using the DagBag would work with DAG serialization?

Reply:
For now I've opted for ensuring the link isn't added if `execution_date_fn` is provided @yuqian90

Reply:
No, looking up the dag in the Dag bag won't work -- the whole goal of enabling serialization is to not need to run the any code in the dag files (or ideally even not need access to the files at all!) from the webserver.

To support when the dag serialization is not enabled/not yet the only option you could do `getattr(operator, 'execution_date_fn')` -- or the other option might be to have the operator store the result of the FN in xcom, and then have the link look up the result from XCom?

Reply:
@ashb I like the idea of using XCom. How would we be able to do an `xcom_pull` in the `get_link` method? It looks like we only have access to the serialized operator and the execution date, but not a task instance

Reply:
@rconroy293 Below is how you could use xcom

https://github.com/apache/airflow/blob/5878e48b787797c54b4138bc0166859232ea08d6/airflow/gcp/operators/bigquery.py#L327-L336

Reply:
Let me know if you want to use Xcom

Reply:
Any updates on the xcom bit @rconroy293 ?

Reply:
@rconroy293  any interests in considering XCom further for implementing the link? If not, I might want to give this a try and put up a PR some time next week.